src/MockGenerator/Program.cs

```diff
@@ -0,0 +1,372 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MockGenerator
+{
+    using System.Diagnostics;
+    using System.IO;
+    using System.Threading;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.MSBuild;
+
+    class Program
+    {
+        static readonly SyntaxTrivia WhitespaceCharacter = SyntaxFactory.Whitespace(" ");
+        static readonly SyntaxTrivia TabCharacter = SyntaxFactory.Whitespace("\t");
+        static readonly SyntaxTrivia NewLineCharacter = SyntaxFactory.Whitespace("\n");
+
+        private static readonly IDictionary<string, ClassDeclarationSyntax> ClassCache;
+        private static readonly IDictionary<string, InterfaceDeclarationSyntax> InterfaceCache;
+
+        static Program()
+        {
+            ClassCache = new Dictionary<string, ClassDeclarationSyntax>();
+            InterfaceCache = new Dictionary<string, InterfaceDeclarationSyntax>();
+        }
+
+        private static void Main()
+        {
+            Run();
+        }
+
+        private static void Run()
+        {
+            var currentDirectory = Environment.CurrentDirectory;
+
+            var workspace = MSBuildWorkspace.Create();
+
+            var solutionRoot = Path.Combine(Path.GetDirectoryName(Path.GetDirectoryName(Path.GetDirectoryName(currentDirectory))), "src");
+            var solution = workspace.OpenSolutionAsync(Path.Combine(solutionRoot, "PInvoke.sln")).Result;
+            var projects = GetProjectsFromSolution(solution);
+            for (var i = 0; i < projects.Length; i++)
+            {
+                var project = projects[i];
+                Console.WriteLine($"Processing project {project.Name}");
+
+                foreach (var file in project.Documents
+                    .Select(x => x.FilePath)
+                    .Where(x => x.EndsWith(".cs")))
+                {
+                    Console.WriteLine($"\tProcessing {Path.GetFileName(file)}");
+                    ProcessSourceCodes(workspace, ref solution, ref project, file);
```

Comment:
Why pass the file path & force `ProcessSourceCodes` to do ore re-do the parsing instead of doing `GetSyntaxTreeAsync` + `GetRootAsync` on the document ?